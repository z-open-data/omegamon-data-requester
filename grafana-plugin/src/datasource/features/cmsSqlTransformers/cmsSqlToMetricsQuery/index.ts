export { detectApplicationAndTable } from './detectApplicationAndTable';
export { parseCmsSqlQuery } from './parseCmsSqlQuery';
export type { CmsSqlParsingProblem, CmsSqlParsingResult } from './parseCmsSqlQuery';
export { buildLinesMap, getTokensPosition } from './SubstringPosition';
export { queries } from './snapshotTestCases';
