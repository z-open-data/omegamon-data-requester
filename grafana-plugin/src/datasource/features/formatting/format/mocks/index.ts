export { ColumnMetadataAndValues } from './ColumnMetadataAndValues';
export { numberEnumColumnMetadataAndValues, stringEnumColumnMetadataAndValues } from './enum';
export {
  columnMdsWithPrintf,
  columnMdsWithoutHexPrintf,
  numberCasesAsNumbersAndAsStrings,
  numberColumnMds,
  brokenNumberCasesAsStrings,
  numberCasesAsNumbers,
} from './number';
export { stringColumnMetadata, stringColumnValues } from './string';
export { invalidItmTimestamps, itmTimestampTestValues, timestampColumnMd, validItmTimestamps } from './timestamp';
