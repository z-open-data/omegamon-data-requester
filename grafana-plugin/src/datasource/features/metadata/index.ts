export { MetadataLoader } from './MetadataLoader';
export { useCurrentTableMetadata, CurrentTableMetadataProvider } from './CurrentTableMetadataProvider';
export { MetadataLoaderProvider, useMetadataLoader } from './MetadataLoaderProvider';
export { useAgentsAndGroups, useApplicationMetadatas, useTableMetadata } from './metadataHooks';
