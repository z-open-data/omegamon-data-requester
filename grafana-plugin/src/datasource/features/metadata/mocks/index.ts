export { mockedAgents } from './mockedAgents';
export { mockedApplicationMetadatas } from './mockedApplicationMetadatas';
export { mockedMetadataLoader } from './mockedMetadataLoader';
export { mockedTableMetadatas, mockedTablesResponse } from './mockedTableMetadatas';
