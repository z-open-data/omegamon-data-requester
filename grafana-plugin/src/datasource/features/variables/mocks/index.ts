export * from './metricsQueryFilters';
export * from './mockedTableMetadatas';
export * from './replaceFnInputOutputMap';
export * from './scopedVars';
export * from './variableModels';
