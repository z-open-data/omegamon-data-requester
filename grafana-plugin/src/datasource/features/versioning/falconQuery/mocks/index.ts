import { AffinityId } from 'public-domain';

import type {
  FalconMetricsQuery0,
  FalconQuery0,
  FalconSituationsQuery0,
  FalconMetricsQuery1,
  FalconSituationsQuery1,
  FalconMetricsQuery2,
  FalconSituationsQuery2,
  FalconQuery_V3,
  FalconQuery_V4,
  FalconQuery_V5,
  FalconQuery_V6,
  FalconQuery_V7,
  FalconQuery_V8,
  FalconQuery_V9,
} from 'datasource/features/versioning/falconQuery';

const commonQueryProperties: Pick<FalconQuery0, 'refId' | 'datasource' | 'hide'> = {
  refId: 'A',
  datasource: {
    type: 'rocketsoftware-omegamon-datasource',
    uid: 'd1317b9c-2ba0-4292-863b-c9c4ae8d2556',
  },
  hide: false,
};

const falconMetricsQuery0: FalconMetricsQuery0 = {
  ...commonQueryProperties,
  queryType: 'regular',
  params: {
    affinityId: '%IBM.STATIC007' as AffinityId,
    applicationCode: 'KM5',
    columns: [
      {
        id: 'ASID',
      },
      {
        id: 'ASTYPE',
      },
    ],
    filter: {},
    first: 0,
    groupBy: [],
    history: false,
    orderBy: [],
    originnodesAndGroups: ['*MVS_SYSTEM'],
    params: [],
    tableId: 'ASCPUUTIL',
  },
};

const falconSituationsQuery0: FalconSituationsQuery0 = {
  ...commonQueryProperties,
  queryType: 'situations',
};

const falconMetricsQuery1: FalconMetricsQuery1 = {
  ...falconMetricsQuery0,
  falconVersion: 1,
};

const falconSituationsQuery1: FalconSituationsQuery1 = {
  ...falconSituationsQuery0,
  falconVersion: 1,
};

const falconMetricsQuery2: FalconMetricsQuery2 = {
  ...falconMetricsQuery0,
  falconVersion: 2,
  queryType: 'regular',
  params: {
    affinityId: '%IBM.STATIC007' as AffinityId,
    applicationCode: 'KM5',
    columns: [
      {
        id: 'ASID',
      },
      {
        id: 'ASTYPE',
      },
    ],
    filter: {},
    first: 0,
    groupBy: [],
    history: false,
    orderBy: [],
    agentsAndGroups: ['*MVS_SYSTEM'],
    parmas: [],
    tableId: 'ASCPUUTIL',
  },
};

const falconSituationsQuery2: FalconSituationsQuery2 = {
  ...falconSituationsQuery0,
  falconVersion: 2,
};

const falconMetricsQuery3: FalconQuery_V3 = {
  ...commonQueryProperties,
  falconVersion: 3,
  queryType: 'regular',
  falconParams: {
    affinityId: '%IBM.STATIC007' as AffinityId,
    applicationCode: 'KM5',
    columns: [
      {
        id: 'ASID',
      },
      {
        id: 'ASTYPE',
      },
    ],
    filter: {},
    first: 0,
    groupBy: [],
    history: false,
    orderBy: [],
    agentsAndGroups: ['*MVS_SYSTEM'],
    parmas: [],
    tableId: 'ASCPUUTIL',
  },
};

const falconSituationsQuery3: FalconQuery_V3 = {
  ...falconSituationsQuery0,
  falconVersion: 3,
};

const falconMetricsQuery4: FalconQuery_V4 = {
  ...commonQueryProperties,
  falconVersion: 4,
  queryType: 'metrics',
  falconParams: {
    affinityId: '%IBM.STATIC007' as AffinityId,
    applicationCode: 'KM5',
    columns: [
      {
        id: 'ASID',
      },
      {
        id: 'ASTYPE',
      },
    ],
    filter: {},
    first: 0,
    groupBy: [],
    history: false,
    orderBy: [],
    agentsAndGroups: ['*MVS_SYSTEM'],
    parmas: [],
    tableId: 'ASCPUUTIL',
  },
};

const falconSituationsQuery4: FalconQuery_V4 = {
  ...falconSituationsQuery0,
  falconVersion: 4,
};

const falconMetricsQuery5: FalconQuery_V5 = {
  ...commonQueryProperties,
  falconVersion: 5,
  queryType: 'metrics',
  falconParams: {
    affinityId: '%IBM.STATIC007' as AffinityId,
    applicationCode: 'KM5',
    columns: [
      {
        id: 'ASID',
      },
      {
        id: 'ASTYPE',
      },
    ],
    filter: {},
    first: 0,
    groupBy: [],
    history: false,
    orderBy: [],
    agentsAndGroups: ['*MVS_SYSTEM'],
    parmas: [],
    tableId: 'ASCPUUTIL',
  },
};

const falconSituationsQuery5: FalconQuery_V5 = {
  ...falconSituationsQuery0,
  falconVersion: 5,
};

const falconMetricsQuery6: FalconQuery_V6 = {
  ...commonQueryProperties,
  falconVersion: 6,
  queryType: 'metrics',
  falconParams: {
    affinityId: '%IBM.STATIC007' as AffinityId,
    applicationCode: 'KM5',
    columns: [
      {
        id: 'ASID',
      },
      {
        id: 'ASTYPE',
      },
    ],
    filter: {},
    first: 0,
    groupBy: [],
    history: false,
    orderBy: [],
    agentsAndGroups: [{ id: 'agentOrGroupName', name: '*MVS_SYSTEM' }],
    parmas: [],
    tableId: 'ASCPUUTIL',
  },
};

const falconSituationsQuery6: FalconQuery_V6 = {
  ...falconSituationsQuery0,
  falconVersion: 6,
};

const falconMetricsQuery7: FalconQuery_V7 = {
  ...commonQueryProperties,
  falconVersion: 7,
  queryType: 'metrics',
  falconParams: {
    affinityId: '%IBM.STATIC007' as AffinityId,
    applicationCode: 'KM5',
    columns: [
      {
        id: 'ASID',
      },
      {
        id: 'ASTYPE',
      },
    ],
    filter: {},
    first: 0,
    groupBy: [],
    history: false,
    orderBy: [],
    agentsAndGroups: [{ id: 'agentOrGroupName', name: '*MVS_SYSTEM' }],
    parmas: [],
    tableId: 'ASCPUUTIL',
  },
};

const falconSituationsQuery7: FalconQuery_V7 = {
  ...falconSituationsQuery0,
  falconVersion: 7,
};

const falconMetricsQuery8: FalconQuery_V8 = {
  ...commonQueryProperties,
  falconVersion: 8,
  queryType: 'metrics',
  falconParams: {
    affinityId: '%IBM.STATIC007' as AffinityId,
    columns: [
      {
        id: 'ASID',
      },
      {
        id: 'ASTYPE',
      },
    ],
    filter: {},
    first: 0,
    groupBy: [],
    history: false,
    orderBy: [],
    agentsAndGroups: [{ id: 'agentOrGroupName', name: '*MVS_SYSTEM' }],
    parmas: [],
    tableId: 'ASCPUUTIL',
  },
};

export const falconSituationsQuery8: FalconQuery_V8 = {
  ...falconSituationsQuery0,
  falconVersion: 8,
};

const falconMetricsQuery9: FalconQuery_V9 = {
  ...commonQueryProperties,
  falconVersion: 9,
  queryType: 'metrics',
  falconParams: {
    affinityId: '%IBM.STATIC007' as AffinityId,
    columns: [
      {
        id: 'ASID',
      },
      {
        id: 'ASTYPE',
      },
    ],
    filter: {},
    first: 0,
    groupBy: [],
    history: false,
    orderBy: [],
    agentsAndGroups: [{ id: 'agentOrGroupName', name: '*MVS_SYSTEM' }],
    parmas: [],
    tableId: 'ASCPUUTIL',
  },
};

export const falconSituationsQuery9: FalconQuery_V9 = {
  ...falconSituationsQuery0,
  falconVersion: 9,
};

type FalconQueryMocks = Array<{ metrics: object; situations: object }>;

export const falconQueryMocks: FalconQueryMocks = [
  { metrics: falconMetricsQuery0, situations: falconSituationsQuery0 },
  { metrics: falconMetricsQuery1, situations: falconSituationsQuery1 },
  { metrics: falconMetricsQuery2, situations: falconSituationsQuery2 },
  { metrics: falconMetricsQuery3, situations: falconSituationsQuery3 },
  { metrics: falconMetricsQuery4, situations: falconSituationsQuery4 },
  { metrics: falconMetricsQuery5, situations: falconSituationsQuery5 },
  { metrics: falconMetricsQuery6, situations: falconSituationsQuery6 },
  { metrics: falconMetricsQuery7, situations: falconSituationsQuery7 },
  { metrics: falconMetricsQuery8, situations: falconSituationsQuery8 },
  { metrics: falconMetricsQuery9, situations: falconSituationsQuery9 },
];
