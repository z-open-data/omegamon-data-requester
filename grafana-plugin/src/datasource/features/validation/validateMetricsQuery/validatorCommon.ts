import { Token } from 'datasource/features/cmsSqlTransformers/cmsSqlToMetricsQuery/tokenizer';

export interface ValidationProblem {
  severity: 'warning' | 'error';
  message: string;
  affectedTokens?: Token[];
}

export const inappropriateWritetimeColumnMessage =
  '"WRITETIME" column can only be used in historical queries but no "HISTORY()" call found in query';
