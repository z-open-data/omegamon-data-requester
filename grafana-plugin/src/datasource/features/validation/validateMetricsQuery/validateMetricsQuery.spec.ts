import { parseCmsSqlQuery } from 'datasource/features/cmsSqlTransformers/cmsSqlToMetricsQuery';
import { mockedMetadataLoader } from 'datasource/features/metadata/mocks';

import { inappropriateWritetimeColumnMessage } from './validatorCommon';

/**
 * During validation of query we check variables too.
 * Current tests are focused on validation check of different part of query but nor variables,
 * So add empty mock of getTemplateSrv to avoid failing tests on this.
 * Validation of variables is handled in separate file on same level: validateVariables.spec.ts
 */
jest.mock('@grafana/runtime', () => ({
  getTemplateSrv: () => ({
    getVariables: () => {
      return [];
    },
  }),
}));

test('Validate correct query', async () => {
  const result = await parseCmsSqlQuery(
    `
      SELECT
        TIMESTAMP, MVSID, SSID, ENCLCPUT, ENCLPER_P_, ENCLPIDX, WRITETIME
      FROM
        KDP.REALTHDA
      HISTORY()
      WHERE
        ((ORIGINNODE = 'DB2plex:DB2plex:Plexview' OR
        ORIGINNODE = 'ICB4:RSD2:DB2' OR
        ORIGINNODE = 'IDS4:DB2plex:DSGROUP' OR
        ORIGINNODE = 'XEDB2:RSD2') );
    `,
    mockedMetadataLoader
  );
  expect(result.queryObject).not.toBeNull();
  expect(result.problems).toHaveLength(4);
  expect(result).toMatchSnapshot();
});

test('Validate columns and filters', async () => {
  const result = await parseCmsSqlQuery(
    `
      SELECT
        TIMESTAMP, MVSID, SSID, ENCLCPUT, ENCLPER_P_, ENCLPIDX, WRITETIME, TESTABCD
      FROM
        KDP.REALTHDA
      WHERE
        ((ORIGINNODE = 'DB2plex:DB2plex:Plexview' OR
        ORIGINNODE = 'ICB4:RSD2:DB2' OR
        ORIGINNODE = 'IDS4:DB2plex:DSGROUP' OR
        ORIGINNODE = 'XEDB2:RSD2') );
    `,
    mockedMetadataLoader
  );
  expect(result.queryObject).not.toBeNull();
  expect(
    result.problems.filter(
      (problem) =>
        problem.stage === 'validator' &&
        problem.severity === 'error' &&
        problem.message === inappropriateWritetimeColumnMessage
    )
  ).toHaveLength(1);
  expect(
    result.problems.filter(
      (problem) =>
        problem.stage === 'validator' &&
        problem.severity === 'warning' &&
        problem.message === 'Column "TESTABCD" not found in table "KDP.REALTHDA"'
    )
  ).toHaveLength(1);
  expect(result.problems).toHaveLength(6);
  expect(result).toMatchSnapshot();
});

test('Validate application code', async () => {
  const result = await parseCmsSqlQuery(
    `
      SELECT
        TIMESTAMP, MVSID, SSID, ENCLCPUT, ENCLPER_P_, ENCLPIDX, WRITETIME
      FROM
        ABC.REALTHDA
      HISTORY()
      WHERE
        ORIGINNODE = 'DB2plex:DB2plex:Plexview' OR
        ORIGINNODE = 'ICB4:RSD2:DB2' OR
        ORIGINNODE = 'IDS4:DB2plex:DSGROUP' OR
        ORIGINNODE = 'XEDB2:RSD2';
    `,
    mockedMetadataLoader
  );
  expect(result).toMatchSnapshot();
});

beforeEach(() => {
  jest.spyOn(console, 'error').mockImplementation(() => {
    void 0;
  });
});

test('Validate table ID', async () => {
  const result = await parseCmsSqlQuery(
    `
      SELECT
        TIMESTAMP, MVSID, SSID, ENCLCPUT, ENCLPER_P_, ENCLPIDX, WRITETIME
      FROM
        KDP.FAKETDA
      HISTORY()
      WHERE
        ORIGINNODE = 'DB2plex:DB2plex:Plexview' OR
        ORIGINNODE = 'ICB4:RSD2:DB2' OR
        ORIGINNODE = 'IDS4:DB2plex:DSGROUP' OR
        ORIGINNODE = 'XEDB2:RSD2';
    `,
    mockedMetadataLoader
  );
  expect(result).toMatchSnapshot();
});

test('Validate GROUP BY', async () => {
  const result = await parseCmsSqlQuery(
    `
      SELECT
        TIMESTAMP, MVSID, SSID, ENCLCPUT, ENCLPER_P_, ENCLPIDX, WRITETIME
      FROM
        KDP.REALTHDA
      HISTORY()
      WHERE
        ORIGINNODE = 'DB2plex:DB2plex:Plexview' OR
        ORIGINNODE = 'ICB4:RSD2:DB2' OR
        ORIGINNODE = 'IDS4:DB2plex:DSGROUP' OR
        ORIGINNODE = 'XEDB2:RSD2'
      GROUP BY
        BLABLA;
    `,
    mockedMetadataLoader
  );
  expect(result).toMatchSnapshot();
});
