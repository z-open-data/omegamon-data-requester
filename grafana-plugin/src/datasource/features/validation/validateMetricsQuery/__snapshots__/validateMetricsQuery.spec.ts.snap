// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Validate GROUP BY 1`] = `
{
  "problems": [
    {
      "message": "ORIGINNODE is handled in SYSTEM.PARMA('NODELIST'), will be moved to this section on Update Query Editor form",
      "position": {
        "endColumn": 47,
        "endLine": 7,
        "globalEnd": 195,
        "globalStart": 156,
        "startColumn": 8,
        "startLine": 7,
      },
      "severity": "informational",
      "stage": "converter",
    },
    {
      "message": "ORIGINNODE is handled in SYSTEM.PARMA('NODELIST'), will be moved to this section on Update Query Editor form",
      "position": {
        "endColumn": 36,
        "endLine": 8,
        "globalEnd": 235,
        "globalStart": 207,
        "startColumn": 8,
        "startLine": 8,
      },
      "severity": "informational",
      "stage": "converter",
    },
    {
      "message": "ORIGINNODE is handled in SYSTEM.PARMA('NODELIST'), will be moved to this section on Update Query Editor form",
      "position": {
        "endColumn": 43,
        "endLine": 9,
        "globalEnd": 282,
        "globalStart": 247,
        "startColumn": 8,
        "startLine": 9,
      },
      "severity": "informational",
      "stage": "converter",
    },
    {
      "message": "ORIGINNODE is handled in SYSTEM.PARMA('NODELIST'), will be moved to this section on Update Query Editor form",
      "position": {
        "endColumn": 33,
        "endLine": 10,
        "globalEnd": 319,
        "globalStart": 294,
        "startColumn": 8,
        "startLine": 10,
      },
      "severity": "informational",
      "stage": "converter",
    },
    {
      "message": "Column "TIMESTAMP" is in SELECT without aggregation function and is not in GROUP BY",
      "position": {
        "endColumn": 17,
        "endLine": 2,
        "globalEnd": 31,
        "globalStart": 22,
        "startColumn": 8,
        "startLine": 2,
      },
      "severity": "error",
      "stage": "validator",
    },
    {
      "message": "Column "MVSID" is in SELECT without aggregation function and is not in GROUP BY",
      "position": {
        "endColumn": 24,
        "endLine": 2,
        "globalEnd": 38,
        "globalStart": 33,
        "startColumn": 19,
        "startLine": 2,
      },
      "severity": "error",
      "stage": "validator",
    },
    {
      "message": "Column "SSID" is in SELECT without aggregation function and is not in GROUP BY",
      "position": {
        "endColumn": 30,
        "endLine": 2,
        "globalEnd": 44,
        "globalStart": 40,
        "startColumn": 26,
        "startLine": 2,
      },
      "severity": "error",
      "stage": "validator",
    },
    {
      "message": "Column "ENCLCPUT" is in SELECT without aggregation function and is not in GROUP BY",
      "position": {
        "endColumn": 40,
        "endLine": 2,
        "globalEnd": 54,
        "globalStart": 46,
        "startColumn": 32,
        "startLine": 2,
      },
      "severity": "error",
      "stage": "validator",
    },
    {
      "message": "Column "ENCLPER_P_" is in SELECT without aggregation function and is not in GROUP BY",
      "position": {
        "endColumn": 52,
        "endLine": 2,
        "globalEnd": 66,
        "globalStart": 56,
        "startColumn": 42,
        "startLine": 2,
      },
      "severity": "error",
      "stage": "validator",
    },
    {
      "message": "Column "ENCLPIDX" is in SELECT without aggregation function and is not in GROUP BY",
      "position": {
        "endColumn": 62,
        "endLine": 2,
        "globalEnd": 76,
        "globalStart": 68,
        "startColumn": 54,
        "startLine": 2,
      },
      "severity": "error",
      "stage": "validator",
    },
    {
      "message": "Column "WRITETIME" is in SELECT without aggregation function and is not in GROUP BY",
      "position": {
        "endColumn": 73,
        "endLine": 2,
        "globalEnd": 87,
        "globalStart": 78,
        "startColumn": 64,
        "startLine": 2,
      },
      "severity": "error",
      "stage": "validator",
    },
    {
      "message": "GROUP BY clause is present but no aggregation functions are found in SELECT clause",
      "position": {
        "endColumn": 73,
        "endLine": 2,
        "globalEnd": 87,
        "globalStart": 7,
        "startColumn": 6,
        "startLine": 1,
      },
      "severity": "error",
      "stage": "validator",
    },
    {
      "message": "Attribute "BLABLA" in attribute group "KDP.REALTHDA" isn't supported by any existing agents and cannot be used in GROUP BY clause.",
      "position": {
        "endColumn": 14,
        "endLine": 12,
        "globalEnd": 349,
        "globalStart": 343,
        "startColumn": 8,
        "startLine": 12,
      },
      "severity": "error",
      "stage": "validator",
    },
  ],
  "queryObject": {
    "affinityId": "%IBM.STATIC017",
    "agentsAndGroups": [
      {
        "id": "agentOrGroupName",
        "name": "DB2plex:DB2plex:Plexview",
      },
      {
        "id": "agentOrGroupName",
        "name": "ICB4:RSD2:DB2",
      },
      {
        "id": "agentOrGroupName",
        "name": "IDS4:DB2plex:DSGROUP",
      },
      {
        "id": "agentOrGroupName",
        "name": "XEDB2:RSD2",
      },
    ],
    "columns": [
      {
        "id": "TIMESTAMP",
      },
      {
        "id": "MVSID",
      },
      {
        "id": "SSID",
      },
      {
        "id": "ENCLCPUT",
      },
      {
        "id": "ENCLPER_P_",
      },
      {
        "id": "ENCLPIDX",
      },
      {
        "id": "WRITETIME",
      },
    ],
    "filter": {
      "nonAggregated": undefined,
    },
    "groupBy": [
      "BLABLA",
    ],
    "history": true,
    "orderBy": [],
    "parmas": [],
    "tableId": "REALTHDA",
  },
  "queryString": "
      SELECT
        TIMESTAMP, MVSID, SSID, ENCLCPUT, ENCLPER_P_, ENCLPIDX, WRITETIME
      FROM
        KDP.REALTHDA
      HISTORY()
      WHERE
        ORIGINNODE = 'DB2plex:DB2plex:Plexview' OR
        ORIGINNODE = 'ICB4:RSD2:DB2' OR
        ORIGINNODE = 'IDS4:DB2plex:DSGROUP' OR
        ORIGINNODE = 'XEDB2:RSD2'
      GROUP BY
        BLABLA;
    ",
}
`;

exports[`Validate application code 1`] = `
{
  "problems": [
    {
      "message": "Specified applicationCode 'ABC' doesn't exist. Available options: KDP, KCP, KM5, O4SRV, APPL",
      "position": {
        "endColumn": 20,
        "endLine": 4,
        "globalEnd": 119,
        "globalStart": 94,
        "startColumn": 6,
        "startLine": 3,
      },
      "severity": "error",
      "stage": "converter",
    },
  ],
  "queryObject": null,
  "queryString": "
      SELECT
        TIMESTAMP, MVSID, SSID, ENCLCPUT, ENCLPER_P_, ENCLPIDX, WRITETIME
      FROM
        ABC.REALTHDA
      HISTORY()
      WHERE
        ORIGINNODE = 'DB2plex:DB2plex:Plexview' OR
        ORIGINNODE = 'ICB4:RSD2:DB2' OR
        ORIGINNODE = 'IDS4:DB2plex:DSGROUP' OR
        ORIGINNODE = 'XEDB2:RSD2';
    ",
}
`;

exports[`Validate columns and filters 1`] = `
{
  "problems": [
    {
      "message": "ORIGINNODE is handled in SYSTEM.PARMA('NODELIST'), will be moved to this section on Update Query Editor form",
      "position": {
        "endColumn": 49,
        "endLine": 6,
        "globalEnd": 191,
        "globalStart": 152,
        "startColumn": 10,
        "startLine": 6,
      },
      "severity": "informational",
      "stage": "converter",
    },
    {
      "message": "ORIGINNODE is handled in SYSTEM.PARMA('NODELIST'), will be moved to this section on Update Query Editor form",
      "position": {
        "endColumn": 36,
        "endLine": 7,
        "globalEnd": 231,
        "globalStart": 203,
        "startColumn": 8,
        "startLine": 7,
      },
      "severity": "informational",
      "stage": "converter",
    },
    {
      "message": "ORIGINNODE is handled in SYSTEM.PARMA('NODELIST'), will be moved to this section on Update Query Editor form",
      "position": {
        "endColumn": 43,
        "endLine": 8,
        "globalEnd": 278,
        "globalStart": 243,
        "startColumn": 8,
        "startLine": 8,
      },
      "severity": "informational",
      "stage": "converter",
    },
    {
      "message": "ORIGINNODE is handled in SYSTEM.PARMA('NODELIST'), will be moved to this section on Update Query Editor form",
      "position": {
        "endColumn": 33,
        "endLine": 9,
        "globalEnd": 315,
        "globalStart": 290,
        "startColumn": 8,
        "startLine": 9,
      },
      "severity": "informational",
      "stage": "converter",
    },
    {
      "message": ""WRITETIME" column can only be used in historical queries but no "HISTORY()" call found in query",
      "position": {
        "endColumn": 73,
        "endLine": 2,
        "globalEnd": 87,
        "globalStart": 78,
        "startColumn": 64,
        "startLine": 2,
      },
      "severity": "error",
      "stage": "validator",
    },
    {
      "message": "Column "TESTABCD" not found in table "KDP.REALTHDA"",
      "position": {
        "endColumn": 83,
        "endLine": 2,
        "globalEnd": 97,
        "globalStart": 89,
        "startColumn": 75,
        "startLine": 2,
      },
      "severity": "warning",
      "stage": "validator",
    },
  ],
  "queryObject": {
    "affinityId": "%IBM.STATIC017",
    "agentsAndGroups": [
      {
        "id": "agentOrGroupName",
        "name": "DB2plex:DB2plex:Plexview",
      },
      {
        "id": "agentOrGroupName",
        "name": "ICB4:RSD2:DB2",
      },
      {
        "id": "agentOrGroupName",
        "name": "IDS4:DB2plex:DSGROUP",
      },
      {
        "id": "agentOrGroupName",
        "name": "XEDB2:RSD2",
      },
    ],
    "columns": [
      {
        "id": "TIMESTAMP",
      },
      {
        "id": "MVSID",
      },
      {
        "id": "SSID",
      },
      {
        "id": "ENCLCPUT",
      },
      {
        "id": "ENCLPER_P_",
      },
      {
        "id": "ENCLPIDX",
      },
      {
        "id": "WRITETIME",
      },
      {
        "id": "TESTABCD",
      },
    ],
    "filter": {
      "nonAggregated": undefined,
    },
    "groupBy": [],
    "history": false,
    "orderBy": [],
    "parmas": [],
    "tableId": "REALTHDA",
  },
  "queryString": "
      SELECT
        TIMESTAMP, MVSID, SSID, ENCLCPUT, ENCLPER_P_, ENCLPIDX, WRITETIME, TESTABCD
      FROM
        KDP.REALTHDA
      WHERE
        ((ORIGINNODE = 'DB2plex:DB2plex:Plexview' OR
        ORIGINNODE = 'ICB4:RSD2:DB2' OR
        ORIGINNODE = 'IDS4:DB2plex:DSGROUP' OR
        ORIGINNODE = 'XEDB2:RSD2') );
    ",
}
`;

exports[`Validate correct query 1`] = `
{
  "problems": [
    {
      "message": "ORIGINNODE is handled in SYSTEM.PARMA('NODELIST'), will be moved to this section on Update Query Editor form",
      "position": {
        "endColumn": 49,
        "endLine": 7,
        "globalEnd": 197,
        "globalStart": 158,
        "startColumn": 10,
        "startLine": 7,
      },
      "severity": "informational",
      "stage": "converter",
    },
    {
      "message": "ORIGINNODE is handled in SYSTEM.PARMA('NODELIST'), will be moved to this section on Update Query Editor form",
      "position": {
        "endColumn": 36,
        "endLine": 8,
        "globalEnd": 237,
        "globalStart": 209,
        "startColumn": 8,
        "startLine": 8,
      },
      "severity": "informational",
      "stage": "converter",
    },
    {
      "message": "ORIGINNODE is handled in SYSTEM.PARMA('NODELIST'), will be moved to this section on Update Query Editor form",
      "position": {
        "endColumn": 43,
        "endLine": 9,
        "globalEnd": 284,
        "globalStart": 249,
        "startColumn": 8,
        "startLine": 9,
      },
      "severity": "informational",
      "stage": "converter",
    },
    {
      "message": "ORIGINNODE is handled in SYSTEM.PARMA('NODELIST'), will be moved to this section on Update Query Editor form",
      "position": {
        "endColumn": 33,
        "endLine": 10,
        "globalEnd": 321,
        "globalStart": 296,
        "startColumn": 8,
        "startLine": 10,
      },
      "severity": "informational",
      "stage": "converter",
    },
  ],
  "queryObject": {
    "affinityId": "%IBM.STATIC017",
    "agentsAndGroups": [
      {
        "id": "agentOrGroupName",
        "name": "DB2plex:DB2plex:Plexview",
      },
      {
        "id": "agentOrGroupName",
        "name": "ICB4:RSD2:DB2",
      },
      {
        "id": "agentOrGroupName",
        "name": "IDS4:DB2plex:DSGROUP",
      },
      {
        "id": "agentOrGroupName",
        "name": "XEDB2:RSD2",
      },
    ],
    "columns": [
      {
        "id": "TIMESTAMP",
      },
      {
        "id": "MVSID",
      },
      {
        "id": "SSID",
      },
      {
        "id": "ENCLCPUT",
      },
      {
        "id": "ENCLPER_P_",
      },
      {
        "id": "ENCLPIDX",
      },
      {
        "id": "WRITETIME",
      },
    ],
    "filter": {
      "nonAggregated": undefined,
    },
    "groupBy": [],
    "history": true,
    "orderBy": [],
    "parmas": [],
    "tableId": "REALTHDA",
  },
  "queryString": "
      SELECT
        TIMESTAMP, MVSID, SSID, ENCLCPUT, ENCLPER_P_, ENCLPIDX, WRITETIME
      FROM
        KDP.REALTHDA
      HISTORY()
      WHERE
        ((ORIGINNODE = 'DB2plex:DB2plex:Plexview' OR
        ORIGINNODE = 'ICB4:RSD2:DB2' OR
        ORIGINNODE = 'IDS4:DB2plex:DSGROUP' OR
        ORIGINNODE = 'XEDB2:RSD2') );
    ",
}
`;

exports[`Validate table ID 1`] = `
{
  "problems": [
    {
      "message": "Specified tableId 'FAKETDA' doesn't exist in application(s) 'DB2 Subsystems' (KDP)",
      "position": {
        "endColumn": 19,
        "endLine": 4,
        "globalEnd": 118,
        "globalStart": 94,
        "startColumn": 6,
        "startLine": 3,
      },
      "severity": "error",
      "stage": "converter",
    },
  ],
  "queryObject": null,
  "queryString": "
      SELECT
        TIMESTAMP, MVSID, SSID, ENCLCPUT, ENCLPER_P_, ENCLPIDX, WRITETIME
      FROM
        KDP.FAKETDA
      HISTORY()
      WHERE
        ORIGINNODE = 'DB2plex:DB2plex:Plexview' OR
        ORIGINNODE = 'ICB4:RSD2:DB2' OR
        ORIGINNODE = 'IDS4:DB2plex:DSGROUP' OR
        ORIGINNODE = 'XEDB2:RSD2';
    ",
}
`;
